### PR TITLE
Ridge regression

### PR DESCRIPTION
**Pull request by VolodymyrOrlov**
Another classical type of linear regression that includes a L2 penalty.

Requirements are summarized in #23

Some resources to understand the code better:

* https://ncss-wpengine.netdna-ssl.com/wp-content/themes/ncss/pdf/Procedures/NCSS/Ridge_Regression.pdf
* https://www.statisticshowto.com/ridge-regression/
* https://www.cs.ubc.ca/~schmidtm/Courses/540-F14/leastSquares.pdf
* http://faculty.cas.usf.edu/mbrannick/regression/regma.htm

### CODE REVIEW

**Review by vadimzaliva — DISMISSED** [review]

File: src/linalg/mod.rs [hunk 1]
```diff
@@ -163,6 +165,32 @@ pub trait BaseVector<T: RealNumber>: Clone + Debug {
...
+
+    /// Computes the arithmetic mean.
+    fn mean(&self) -> T {
+        let n = self.len();
+        let mut mean = T::zero();
+
+        for i in 0..n {
+            mean += self.get(i);
+        }
+        mean / T::from_usize(n).unwrap()
```
Comment (morenol) [inline src/linalg/mod.rs:177]: Couldn't this be changed to?

self.sum() / T::from_usize(n).unwrap()

File: src/linalg/mod.rs [hunk 2]
```diff
@@ -163,6 +165,32 @@ pub trait BaseVector<T: RealNumber>: Clone + Debug {
...
+    }
+    /// Computes the standard deviation.
+    fn std(&self) -> T {
```
Comment (morenol) [inline src/linalg/mod.rs:180]: Does it make sense to have something to calculate the variance? and use here `self.variance().sqrt()`?

File: src/linear/ridge_regression.rs [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+    /// * `y` - target values
+    /// * `parameters` - other parameters, use `Default::default()` to set parameters to default values.
+    pub fn fit(
+        x: &M,
+        y: &M::RowVector,
+        parameters: RidgeRegressionParameters<T>,
+    ) -> Result<RidgeRegression<T, M>, Failed> {
+        //w = inv(X^t X + alpha*Id) * X.T y
+
+        let (n, p) = x.shape();
```
Comment (morenol) [inline src/linear/ridge_regression.rs:129]: Maybe we should check here that the size of y is `n`

File: src/linear/ridge_regression.rs [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+    pub fn predict(&self, x: &M) -> Result<M::RowVector, Failed> {
+        let (nrows, _) = x.shape();
+        let mut y_hat = x.matmul(&self.coefficients);
+        y_hat.add_mut(&M::fill(nrows, 1, self.intercept));
+        Ok(y_hat.transpose().to_row_vector())
+    }
+
+    /// Get estimates regression coefficients
+    pub fn coefficients(&self) -> M {
+        self.coefficients.clone()
```
Comment (morenol) [inline src/linear/ridge_regression.rs:220]: Do we need to clone here?

Comment (VolodymyrOrlov) [reply src/linear/ridge_regression.rs:220]: with the current interface we do need to clone. If I decide to change this interface I should also change interfaces of `LinearRegression` and `LogisticRegression` as well for consistency reasons. 

I agree that returning a reference here is an overkill

**Comment by morenol** [comment]
@VolodymyrOrlov Could you ensure that the changes made in this PR would not create new clippy errors if we rebase this changes on https://github.com/smartcorelib/smartcore/pull/24? Or maybe we just need to merge that PR and rebase this one and see if there are errors?

**Comment by VolodymyrOrlov** [comment]
> @VolodymyrOrlov Could you ensure that the changes made in this PR won't create new clippy errors if we rebase this changes on #24? Or maybe we just need to merge that PR and rebase this one and see if there are errors?

This is a good idea. Shell we merge #24 if you don't plan any more changes to it?

**Comment by morenol** [comment]
No

> > @VolodymyrOrlov Could you ensure that the changes made in this PR won't create new clippy errors if we rebase this changes on #24? Or maybe we just need to merge that PR and rebase this one and see if there are errors?
> 
> This is a good idea. Shell we merge #24 if you don't plan any more changes to it?

Yes, I am going to merge that. I plan to fix the other warnings in  a follow up PR

**Review by morenol — APPROVED** [review]

**Pull request merged**